src/mpi/coll/reduce/reduce.c

```diff
@@ -321,6 +334,94 @@ int MPIR_Reduce_impl(const void *sendbuf, void *recvbuf, int count,
     goto fn_exit;
 }
 
+#undef FUNCNAME
+#define FUNCNAME MPIR_Reduce_decision_tree
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
+int MPIR_Reduce_decision_tree(const void *sendbuf, void *recvbuf, int count, MPI_Datatype datatype,
+                              MPI_Op op, int root, MPIR_Comm * comm_ptr, MPIR_Errflag_t * errflag)
+{
+    int mpi_errno = MPI_SUCCESS;
+    size_t size;
+    cJSON *device_info;
+
+    /* TODO - Calculate size of buffer */
+
+    mpi_errno = MPIU_Decision_tree_traverse("reduce", comm_ptr, size, datatype, coll_info);
+
+    if (unlikely(mpi_errno != MPI_SUCCESS))
+        MPIR_ERR_POP(mpi_errno);
+    else {
+        cJSON *mpir = cJSON_GetObjectItemCaseSensitive(coll_info, "mpir");
+
+        /* If the JSON tree is using an MPIR-level collective, parse that information here. */
+        if (likely(cJSON_IsString(mpir) && 0 != strcmp(mpir->valuestring, "device"))) {
+            /* TODO - Combine MPIR_Reduce_intra_algo_t and MPIR_Reduce_inter_algo_t */
+            MPIR_Reduce_algo_t algo_choice;
+
+            /* TODO - Separate out all of the CVAR parsing from the MPII_Coll_init function so it
+             * can be called here with the string in the JSON object. */
+            MPIR_Coll_parse_name_reduce(mpir->valuestring, &algo_choice);
+
+            /* Call the algorithm chosen by the decision tree */
+            /* TODO - Separate out this part from this function and the function above to only write
+             * it once. */
+            switch (algo_choice) {
```

Comment:
This is a _call phase we have after _select which returns container. So switch works with coll_id and we propagate collective operation parameters to collective operation